File: ios/Stopwatch/infra/di/impl/ApplicationContainerFactoryImpl.swift
class ApplicationContainerFactoryImpl: ApplicationContainerFactory {
    typealias Stores = ApplicationStores<StoresStopwatch>
    typealias StoresStopwatch = MutableStateStoreImpl<StopwatchState, EventEmitterImpl<StopwatchState>>
    typealias Repositories = ApplicationRepositories
    typealias Services = ApplicationServices<ServicesTimer>
    typealias ServicesTimer = TimerServiceImpl<EventEmitterImpl<TimerState>>
    typealias UseCases = ApplicationUseCases
    typealias Presentation = ApplicationPresentation<PresentationHVMF, PresentationLVMF, PresentationSN>
    typealias PresentationHVMF = HomeViewModelFactoryImpl<StoresStopwatch, PresentationSN>
    typealias PresentationLVMF = LapsViewModelFactoryImpl<StoresStopwatch, PresentationSN>
    typealias PresentationSN = StackNavigatorImpl<EventEmitterImpl<[StackNavigatorPath]>>
    typealias Container = ApplicationContainer<
        Stores,
        StoresStopwatch,
        Repositories,
        Services,
        ServicesTimer,
        Presentation,
        PresentationHVMF,
        PresentationLVMF,
        PresentationSN
    >
    
    func make() -> Container {
        let stores = ApplicationStores(MutableStateStoreImpl(StopwatchState(), EventEmitterImpl<StopwatchState>()))
        let repositories = ApplicationRepositories(StopwatchRepositoryImpl(StopwatchDataSourceFileManager()))
        let services = ApplicationServices(TimerServiceImpl(EventEmitterImpl()))
        let useCases = ApplicationUseCases(
            StartStopwatchUseCaseImpl(stores.stopwatch, services.timer),
            PauseStopwatchUseCaseImpl(stores.stopwatch, services.timer),
            ResetStopwatchUseCaseImpl(stores.stopwatch, services.timer),
            NewLapUseCaseImpl(stores.stopwatch),
            UpdateStopwatchTimeAndLapsUseCaseImpl(stores.stopwatch),
            SaveStopwatchStateUseCaseImpl(stores.stopwatch, repositories.stopwatch),
            RestoreStopwatchStateUseCaseImpl(stores.stopwatch, repositories.stopwatch, services.timer)
        )
        
        let viewTimeMapper = ViewTimeMapperImpl()
        let stringTimeMapper = StringTimeMapperImpl(viewTimeMapper)
        let stackNavigator = StackNavigatorImpl(stack: [.home], eventEmitter: EventEmitterImpl())
        let homeViewModelFactory = HomeViewModelFactoryImpl(
            stores.stopwatch,
            useCases.startStopwatch,
            useCases.pauseStopwatch,
            useCases.resetStopwatch,
            useCases.newLap,
            viewTimeMapper,
            stringTimeMapper,
            stackNavigator
        )
        let lapsViewModelFactory = LapsViewModelFactoryImpl(
            stores.stopwatch,
            useCases.startStopwatch,
            useCases.newLap,
            stringTimeMapper,
            stackNavigator
        )

        let presentation = ApplicationPresentation(
            homeViewModelFactory,
            lapsViewModelFactory,
            stackNavigator
        )

        let container = ApplicationContainer(
            stores,
            repositories,
            services,
            useCases,
            presentation
        )
        
        return container
    }
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/platform/di/impl/ApplicationContainerFactoryImpl.kt
package com.rafaeltmbr.stopwatch.platform.di.impl

import android.content.Context
import com.rafaeltmbr.stopwatch.core.data.repositories.impl.StopwatchRepositoryImpl
import com.rafaeltmbr.stopwatch.core.data.stores.impl.MutableStateStoreImpl
import com.rafaeltmbr.stopwatch.core.entities.StopwatchState
import com.rafaeltmbr.stopwatch.core.services.impl.LoggingServiceImpl
import com.rafaeltmbr.stopwatch.core.services.impl.TimerServiceImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.NewLapUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.PauseStopwatchUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.ResetStopwatchUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.RestoreStopwatchStateUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.SaveStopwatchStateUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.StartStopwatchUseCaseImpl
import com.rafaeltmbr.stopwatch.core.use_cases.impl.UpdateStopwatchTimeUseCaseImpl
import com.rafaeltmbr.stopwatch.core.utils.impl.CalculateLapsStatusesImpl
import com.rafaeltmbr.stopwatch.platform.data.room.StopwatchDatabase
import com.rafaeltmbr.stopwatch.platform.data.room.data_sources.RoomStopwatchDataSourceAdapter
import com.rafaeltmbr.stopwatch.platform.di.ApplicationContainer
import com.rafaeltmbr.stopwatch.platform.di.ApplicationContainerFactory
import com.rafaeltmbr.stopwatch.platform.presentation.compose.navigation.StackNavigatorImpl
import com.rafaeltmbr.stopwatch.platform.presentation.mappers.impl.StringTimeMapperImpl
import com.rafaeltmbr.stopwatch.platform.presentation.mappers.impl.ViewTimeMapperImpl
import com.rafaeltmbr.stopwatch.platform.presentation.navigation.StackNavigator
import com.rafaeltmbr.stopwatch.platform.services.external_resources.AndroidPlatformLoggerAdapter

class ApplicationApplicationContainerFactoryImpl(private val context: Context) :
    ApplicationContainerFactory {
    override fun make(): ApplicationContainer {
        val database = StopwatchDatabase.getInstance(context)

        val data = ApplicationContainer.Data(
            MutableStateStoreImpl(StopwatchState()),
            StopwatchRepositoryImpl(
                RoomStopwatchDataSourceAdapter(
                    database.stopwatchStateDao(),
                    database.lapsDao()
                )
            )
        )

        val services = ApplicationContainer.Services(
            TimerServiceImpl(),
            LoggingServiceImpl(AndroidPlatformLoggerAdapter())
        )

        val calculateLapsStatuses = CalculateLapsStatusesImpl()
        val useCases = ApplicationContainer.UseCases(
            NewLapUseCaseImpl(data.stopwatchStore, calculateLapsStatuses),
            PauseStopwatchUseCaseImpl(data.stopwatchStore, services.timer),
            ResetStopwatchUseCaseImpl(data.stopwatchStore, services.timer),
            RestoreStopwatchStateUseCaseImpl(
                data.stopwatchStore,
                data.stopwatchRepository,
                services.timer,
            ),
            SaveStopwatchStateUseCaseImpl(
                data.stopwatchStore,
                data.stopwatchRepository,
            ),
            StartStopwatchUseCaseImpl(data.stopwatchStore, services.timer),
            UpdateStopwatchTimeUseCaseImpl(data.stopwatchStore),
        )

        val stackNavigator = StackNavigatorImpl(listOf(StackNavigator.Screen.Home))
        val viewTimeMapper = ViewTimeMapperImpl()
        val stringTimeMapper = StringTimeMapperImpl(ViewTimeMapperImpl())

        val presentation = ApplicationContainer.Presentation(
            stackNavigator,
            HomeViewModelFactoryImpl(
                data.stopwatchStore,
                stackNavigator,
                useCases.startStopwatch,
                useCases.pauseStopwatch,
                useCases.resetStopwatch,
                useCases.newLap,
                services.logging,
                viewTimeMapper,
                stringTimeMapper
            ),
            LapsViewModelFactoryImpl(
                data.stopwatchStore,
                stackNavigator,
                useCases.startStopwatch,
                useCases.newLap,
                services.logging,
                stringTimeMapper
            ),
        )

        return ApplicationContainer(
            data,
            services,
            useCases,
            presentation
        )
    }
}